### PR TITLE
Create Database.cpp file

### PR DESCRIPTION
**Pull request by firewings2515**

### CODE REVIEW

**Review by YuanWeiChen16 — APPROVED** [review]
viewed

File: src/Database.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+bool Database::DeleteSURL() {
+  Date now = boost::gregorian::day_clock::local_day();
+  bool delete_flag = false;
+
+  std::unordered_map<std::string, std::unique_ptr<SURL>>::iterator it;
+  it = lut.begin();
+  while (it != lut.end()) {
+    boost::gregorian::date_duration dur =
+        now - it->second.get()->GetExpireTime();
+    if (dur.days() >= 30) {
```
Comment (john0312) [inline src/Database.cpp:86]: Please make this a constant. i.e. `constexpr int kExpireTimeInDays`;

File: src/Database.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      it = lut.erase(it);
```
Comment (john0312) [inline src/Database.cpp:87]: You might need to lock write_lock.

File: src/Database.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+  } else {
+    // if short url does not in the lut, insert it to the lut
+    std::unique_ptr<SURL> surl(new SURL(raw_url, short_url));
+    InsertSURL(short_url, std::move(surl));
+    return short_url;
+  }
+}
+std::string Database::ReadSURL(const std::string &surl) const {
+  // if the surl is not in the lut, return "", else return the raw url.
```
Comment (john0312) [inline src/Database.cpp:44]: Consider locking `write_lock` here as well, unless you are sure that the lut update is atomic. i.e. If we read while we're writing, could it cause any error or incorrect read?

File: src/Database.h [hunk 4]
```diff
@@ -15,71 +15,74 @@ found in the LICENSE file. See the AUTHORS file for names of contributors.
...
+  std::string CreateSURL(const std::string &raw_url);
+
+  // Read the lut whether URL exists. Return rawURL if SURL exists. Return "" if
+  // SURL doesn't exist.
+  std::string ReadSURL(const std::string &surl) const;
+
+private:
+  // lock the mutex when it is inserting.
+  // There is only one inserting in the same time.
+  bool InsertSURL(std::string short_url, std::unique_ptr<SURL> surl);
```
Comment (john0312) [inline src/Database.h:64]: Please use const for input argument.

File: src/Database.h [hunk 5]
```diff
@@ -15,71 +15,74 @@ found in the LICENSE file. See the AUTHORS file for names of contributors.
...
+
+  // If there is a rawURL which encoding to a SURL that exists in LUT,
+  // Update its expire time and don't need to create new one in LUT.
+  bool UpdateSURL(std::unique_ptr<SURL> surl, Date expire);
```
Comment (john0312) [inline src/Database.h:68]: Please use const for input argument `expire`

File: src/SURLCreator.cpp [hunk 6]
```diff
@@ -45,7 +47,7 @@ std::string SURLCreator::CreateSURL(const std::string &raw_url, int length) {
   }
 
   // extract a string with a length of 6 ~ 10, depending on the parameter.
-  surl_key = surl_key.substr(0, length);
+  surl_key = surl_key.substr(10 - length, length);
```
Comment (john0312) [inline src/SURLCreator.cpp:50]: Please make 10 a constant and check for input `length` correctness.

**Review by john0312 — COMMENTED** [review]
Consider adding a unit test in this PR or another PR.

**Comment by john0312** [comment]
Note: If you don't have time to finish the unit test, please let me know as well. I'll approve this PR first in that case.

**Comment by firewings2515** [comment]
I add a unit test in another pr.

**Comment by john0312** [comment]
Still LGTM.

**Pull request merged**